src/dotty/tools/dotc/core/TypeApplications.scala

```diff
@@ -326,112 +306,57 @@ class TypeApplications(val self: Type) extends AnyVal {
         if (params.exists(_.name == self.refinedName)) parent1
         else self.derivedRefinedType(parent1, self.refinedName, self.refinedInfo)
       case self: TypeProxy =>
-        self.underlying.widenToNamedTypeParams(params)
+        self.superType.widenToNamedTypeParams(params)
       case self: AndOrType =>
         self.derivedAndOrType(
           self.tp1.widenToNamedTypeParams(params), self.tp2.widenToNamedTypeParams(params))
     }
   }
 
-  /** The Lambda trait underlying a type lambda */
-  def LambdaTrait(implicit ctx: Context): Symbol = self.stripTypeVar match {
-    case RefinedType(parent, tpnme.hkApply) =>
-      val sym = self.classSymbol
-      if (sym.isLambdaTrait) sym else NoSymbol
-    case TypeBounds(lo, hi) => hi.LambdaTrait
-    case _ => NoSymbol
-  }
-
-  /** Is receiver type higher-kinded (i.e. of kind != "*")? */
+  /** Is self type higher-kinded (i.e. of kind != "*")? */
   def isHK(implicit ctx: Context): Boolean = self.dealias match {
     case self: TypeRef => self.info.isHK
-    case RefinedType(_, name) => name == tpnme.hkApply
-    case TypeBounds(_, hi) => hi.isHK
+    case self: RefinedType => false
+    case self: TypeLambda => true
+    case self: HKApply => false
+    case self: SingletonType => false
+    case self: TypeVar => self.origin.isHK // discrepancy with typeParams, why?
```

Comment:
Good question. Using `origin` avoids having to set `ephemeral` to `true` in some cases, and is enough to know whether a type is higher-kinded or not. But you can't use `origin` to get the correct type parameters of any type.